src/uct/base/uct_worker.c

```diff
@@ -70,50 +60,31 @@ void uct_worker_progress_unregister(uct_worker_h worker,
     UCS_ASYNC_BLOCK(worker->async);
     ucs_assert(prog->refcount > 0);
     if (--prog->refcount == 0) {
-        ucs_callbackq_remove(&worker->progress_q, prog->cb, prog->arg);
+        ucs_callbackq_remove(&worker->progress_q, prog->id);
+        prog->id = UCS_CALLBACKQ_ID_NULL;
     }
     UCS_ASYNC_UNBLOCK(worker->async);
 }
 
-static void uct_worker_slowpath_proxy(ucs_callbackq_slow_elem_t *self)
-{
-    uct_worker_slowpath_elem_t *elem = ucs_derived_of(self, uct_worker_slowpath_elem_t);
-    elem->cb(elem->arg);
-}
-
 void uct_worker_progress_register_safe(uct_worker_h worker, ucs_callback_t func,
                                        void *arg, unsigned flags,
                                        uct_worker_cb_id_t *id_p)
 {
-    uct_worker_slowpath_elem_t *elem;
-
-    if (*id_p == NULL) {
+    if (*id_p == UCS_CALLBACKQ_ID_NULL) {
```

Comment:
according to the API, which was defined by the usage of slowpath progress that we have today, we mostly want to register the callback if not yet registered, and remove it (completely) if it was. for example, we had the "cbq_elem_on" flag in multiple places (uct/mm, cm, ucp_ep..).